Introduce a schema/app store for all our app types (#1348)
platformatic/platformatic: docs/guides/packaging-an-application-as-a-module.md

# Packaging a Platformatic Application as a module

[Platformatic Service](/docs/reference/db/introduction.md) and [Platformatic DB](/docs/reference/db/introduction.md)

> offer a good starting point to create new applications. However most developers or organization might want to

Reviewer: ```suggestion
offer a good starting point to create new applications. However, most developers or organizations might want to
```

> create reusable services or applications build on top of Platformatic.

Reviewer: ```suggestion
create reusable services or applications built on top of Platformatic.
```

> This is useful to publish the application on the public npm registry (or a private one!), including buildig your own CLI,

Reviewer: ```suggestion
This is useful to publish the application on the public npm registry (or a private one!), including building your own CLI,
```

/**  @type {import('fastify').FastifyPluginAsync<{}>} */
async function foo (app) {
  const text = app.platformatic.config.foo.text
  app.get('/foo', async (request, reply) => {
    return text
  })
}

foo.configType = 'foo'

// The schema for our configuration file
foo.schema = {
  $id: 'https://example.com/schemas/foo.json',
  title: 'Foo Service',
  type: 'object',
  properties: {
    server: schema.server,
    plugins: schema.plugins,
    metrics: schema.metrics,
    watch: {
      anyOf: [schema.watch, {
        type: 'boolean'
      }, {
        type: 'string'
      }]
    },
    $schema: {
      type: 'string'
    },
    module: {
      type: 'string'
    },
    foo: {
      type: 'object',
      properties: {
        text: {
          type: 'string'
        }
      },
      required: ['text']
    }
  },
  additionalProperties: false,
  required: ['server']
}

// The configuration for the ConfigManager
foo.configManagerConfig = {
  schema: foo.schema,
  envWhitelist: ['PORT', 'HOSTNAME'],
  allowToWatch: ['.env'],
  schemaOptions: {
    useDefaults: true,
    coerceTypes: true,
    allErrors: true,
    strict: false
  }
}

module.exports = foo
```

> Note that the `$id` property of the schema identify the module in our system,

Reviewer: ```suggestion
Note that the `$id` property of the schema identifies the module in our system,
```

allowing us to retrieve the schema correctly.

> Given that this schema is what will be used for the configuration file,
> it's better if this JSON schema is actually published there, so that
> the VSCode language server will provide autocompletion in that case.

Reviewer: ```suggestion
It is recommended, but not required, that the JSON schema is actually
published in this location. Doing so allows tooling such as the VSCode
language server to provide autocompletion.
```
